### PR TITLE
feat(networking): HTTP proxy support

### LINKED ISSUE #6401: HTTP proxy support
Requested by user in discord: https://discord.com/channels/742820443487993987/746070591097798688/808697724547104815

I think we can take a general approach, similar to other HTTP configuration, to allow for the use of an HTTP proxy with any component that makes HTTP requests. We could allow specification of the proxy both at the global and component level.

Suggested config:

Global:

```toml
[proxy]
http = "http://myproxy:3128"
https = "http://myproxy:3128"
no_proxy = ["localhost", "192.168.1.1"]
```

Component:

```toml
[sinks.http]
type = "http"
proxy.http = "http://myproxy:3128"
proxy.https = "http://myproxy:3128"
proxy.no_proxy = ["localhost", "192.168.1.1"]
```

Additionally we should read from the following env vars:

```
HTTP_PROXY
HTTPS_PROXY
NO_PROXY
http_proxy
https_proxy
no_proxy
```

These are commonly supported env vars for this purpose (for example `curl` respects them). `no_proxy` should accept a comma delimited list.

Ref: https://github.com/timberio/vector/issues/4534

### PR DESCRIPTION
**Pull request by jdrouet**
closes #6401

- creates a global proxy config
- add proxy configs in every sources/sinks using `HttpClient`
- source proxy config from environment variables
- merge local config with global config (example: ENV < Global < ES config)

## TODO

- [x] Add `proxy.enabled` flag
- [x] `Regarding rusoto it can use a custom hyper::client::Builder so most of this PR can be reused easily for requests to AWS.`
- [x] Regarding tests, we should probably ensure that IPv6 addresses are working and whether they need to be enclosed in brackets, especially if mixed with IPv4/hostnames in the no_proxy env var. (Done in `no-proxy` crate)
- [x] Add more unit tests
- [x] Add an integration test

### CODE REVIEW

File: src/config/proxy.rs
Comment (jszwedko) [inline src/config/proxy.rs:58]:
```suggestion
            https: from_env("HTTPS_PROXY"),
```

File: src/config/proxy.rs
Comment (jszwedko) [inline src/config/proxy.rs:30]: I've observed that different clients implement `no_proxy` differently. I think what you have here is good, but what would you think of expanding it to cover a few more cases as suggested by https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/ ?

> Use * to match all hosts.
> Strip leading dots (.) and match against domain suffixes.
> Support CIDR block matching.

I'm noting that they also recommend checking for lowercased versions before uppercased. I don't have a strong opinion on that.

**Review by jszwedko — COMMENTED** [review]
Nice work!

In addition to the couple of comments I left:

* Is it possible to _unset_ the proxy configuration for a given component? I guess you could just do `no_proxy="*"` (at least if the other suggestion about wildcards is implemented), but I'm wondering if we could make it more straight-forward with something like `proxy.enabled = false|true` so that individual components could easily disable the proxy as needed.
* We'll need some docs. This configuration would be shared across all http-based components so we could DRY it up by creating one cue block they could all reference.
* Should the `aws_*` sources/sinks support this as well? I realize they have their own SDK, but I think we can inject the HTTP client it uses.
* I'm wondering about tests for this, but I realize it'd could be a little much to test every component individually. I'll see if I think of anything better.

**Comment by prognant** [comment]
Nice work, I also think that we should support both lower case/upper case env var. Regarding rusoto it can [use](https://docs.rs/rusoto_core/0.47.0/rusoto_core/request/struct.HttpClient.html#method.from_builder) a custom `hyper::client::Builder` so most of this PR can be reused easily for requests to AWS. Regarding tests, we should probably ensure that IPv6 addresses are working and whether they need to be enclosed in brackets, especially if mixed with IPv4/hostnames in the `no_proxy` env var.

Just for reference the `no_proxy` behaviour was recently revisited in the agent DataDog/datadog-agent#6897 to support suffix matching, so I guess this one matters to some users.

File: src/config/proxy.rs
Comment (jdrouet) [reply src/config/proxy.rs:30]: Considering the "complexity" of this `NO_PROXY` proposed by the people from gitlab (good article btw) and the fact that it's propably going to be reimplemented by someone else, would it be worth making a separate crate for it? a `no-proxy-parser` of some sort? Or even a `proxy-config-parser` 🤔

Comment (jszwedko) [reply src/config/proxy.rs:30]: Just noting that we talked outside and a separate crate for handling no_proxy matching seems reasonable given the complexity.

**Comment by jdrouet** [comment]
> I'm wondering about tests for this, but I realize it'd could be a little much to test every component individually. I'll see if I think of anything better.

@jszwedko considering they all use the same `HttpClient` implementation, we could _just_ write an integration test that pushes data through a proxy

**Comment by binarylogic** [comment]
Just to chime in regarding testing - we've been bit in the past with changes like this. Where we add new configuration parameters, without testing them, and realize they don't work as expected. Defaults aren't respected, global defaults aren't applied properly (all or nothing), and so on. See #8229.

I'd really like to solve how we test parsing of configuration values so we don't keep adding to the pile of problems. For example, do we know if individual proxy fields can be overridden on a component-level while still preserving the global defaults? If that's the behavior we want, we should test for that.

**Review by jszwedko — COMMENTED** [review]
This is coming together, thanks @jdrouet ! I left a few more comments, but nothing too major.

To @binarylogic's point, we still aren't really testing that each component gets configured directly given the proxy configuration, but I'm not seeing an easy way to do that since this happens when the component is built and results in a custom future. I think maybe we'd need to change the components to have a method to expose their HTTP client which we could then assert on. Given we lack tests for other similar things like merging `request` defaults for `Tower`, I don't think it should necessarily block this PR, but I am curious if anyone has clever ideas.

I realized you backed out the integration test, but I might actually still like to see one integration test for the `http` sink component that checks that things are proxied correctly. I realize this is the domain of `hyper` and `hyper-proxy`, but one high-level integration test should give us some more confidence. Apologies, I realize there's been some back-and-forth around that.

File: src/config/proxy.rs
Comment (jdrouet) [reply src/config/proxy.rs:93]: good point

File: docs/cue/reference/components/sources.cue
Comment (jszwedko) [inline docs/cue/reference/components/sources.cue:121]: Was there a source that is "receive" that uses the proxy? I think the proxy might only be relevant for "collect" sources which initiate requests.

File: src/config/proxy.rs
Comment (jszwedko) [inline src/config/proxy.rs:32]:
```suggestion
                // only intercept those that don't match
```

File: src/kubernetes/client/mod.rs
Comment (jszwedko) [inline src/kubernetes/client/mod.rs:66]: Is this right? How is the global config file proxy config injected here? Does `config.proxy` already take the global settings into account?

File: src/sinks/aws_cloudwatch_logs/mod.rs
Comment (jszwedko) [inline src/sinks/aws_cloudwatch_logs/mod.rs:174]: It seems like you missed this one when fixing the AWS components to take the global proxy settings into account.

File: src/transforms/aws_ec2_metadata.rs
Comment (jszwedko) [inline src/transforms/aws_ec2_metadata.rs:85]: I think we actually need to add this config to the docs for this transform still.

File: docs/cue/reference/components.cue
Comment (jszwedko) [reply docs/cue/reference/components.cue:717]: I think this comment is still relevant.

Do we allow configuring `no_proxy` as a list in the config? It doesn't seem like it though I think we should.

We should also still document exactly how `no_proxy` matches.

**Review by jszwedko — COMMENTED** [review]
Nice, thanks for the updates!

I left a few more comments.

I'd still like to see an integration test for this with the `http` sink where we verify everything works as expected. Is that something you think would be doable? I know you had previously started one with the `nginx_metrics` source that could be used as a base.

File: src/kubernetes/client/mod.rs
Comment (jdrouet) [reply src/kubernetes/client/mod.rs:66]: That's a good point. I went too fast on these.

**Comment by jdrouet** [comment]
> I'd still like to see an integration test for this with the http sink where we verify everything works as expected. Is that something you think would be doable? I know you had previously started one with the nginx_metrics source that could be used as a base.

I agree this would be interesting. I've had some weird issues trying to do that. I've started the following `docker-compose.yml` locally

```yml
version: '2'

services:
  squid:
    image: babim/squid
    environment:
      - PUBLIC=true
    ports:
      - 3128:3128
    networks:
      - backend

  nginx:
    image: nginx:1.19.4
    networks:
      - backend
    volumes:
      - "${VECTOR_PATH}/tests/data/nginx/:/etc/nginx:ro"
    ports:
      - 8010:8000

networks:
  backend: {}
```

With the following `vector.toml`

```toml
[sources.nginx]
type = "nginx_metrics"
endpoints = ["http://nginx:8000/basic_status"]
proxy.http = "http://localhost:3128"

[sinks.console]
type = "console"
inputs = ["nginx"]
encoding.codec = "json"
```

And it works like a charm:
- vector goes through the proxy (because configured to point on `8000` which is not exposed by the `nginx` container)
- the `nginx` ip is resolved in the proxy container

But when I modify this to make it work with the integrations tests, it starts being weird.

```bash
# in scripts/setup_integration/nginx_integration_env.sh
start_docker () {
  docker network create vector-test-integration-nginx
  docker run -d --network=vector-test-integration-nginx -p 3128:3128 --name vector_squid \
        -e PUBLIC=true babim/squid
  docker run -d --network=vector-test-integration-nginx -p 8010:8000 --name vector_nginx \
        -v "$(pwd)"/tests/data/nginx/:/etc/nginx:ro nginx:1.19.4
}
```

And when I do this, when running tests, it appears that the request doesn't even leave from Vector because it cannot resolve `nginx` IP. Although on both cases it runs on my host, not in the container so the `nginx` IP should be resolvable.

Anyway, I'll try to dig more into this.

**Comment by jdrouet** [comment]
integration test working: https://github.com/timberio/vector/runs/3179767637?check_suite_focus=true

File: docs/cue/reference/components.cue
Comment (jszwedko) [inline docs/cue/reference/components.cue:728]:
```suggestion
						description: "List of hosts to avoid proxying."
```

File: src/sources/nginx_metrics/mod.rs
Comment (jszwedko) [inline src/sources/nginx_metrics/mod.rs:326]: I might just add a comment here that this is the test we have that verifies simple proxy support in Vector. Users might otherwise think this test exists for all components. Maybe something like:

```
// This integration test verifies that proxy support is wired up correctly in Vector
// It is the only test of its kind
```

**Review by jszwedko — COMMENTED** [review]
Awesome, this looks good to me. Thanks for re-adding the integration test.

One last thought that occurred to me is that, since `no_proxy` configs are merged rather than replaced when merging, there is no way to completely override the `no_proxy` config at the component level. What would you think of, instead of merging, we simply replaced `no_proxy` config? I think this would be consistent with how other configuration overrides are supplied. The trade-off is that users will need to specify the full set when overriding rather than just suppling additional hosts.

Also, I realized we should ensure these new environment variables show up on https://deploy-preview-8205--vector-project.netlify.app/docs/reference/cli/#environment-variables . I think we can do that like this: https://github.com/timberio/vector/blob/4c7eaf4c8f72d897f30f32fbf553142066691e9b/docs/cue/reference/components/aws.cue#L73-L89

Maybe adding them `component.cue` since they are so globally applicable.

**Comment by jdrouet** [comment]
test `sources::internal_logs::tests::receives_logs` failed (not related)

**Review by jszwedko — APPROVED** [review]
Awesome, thanks @jdrouet ! I'll let Bruce finish his review.

File: src/config/proxy.rs
Comment (bruceg) [inline src/config/proxy.rs:19]: Could `scheme` validly also be `None` here?

File: src/config/proxy.rs
Comment (bruceg) [inline src/config/proxy.rs:31]: This could be done with adapters, but I'm on the fence if it improves readability. I think we *should* be doing the host match first before constructing the URL to compare it in any case.
```suggestion
                let matches = host.map(|host| {
                    self.0.matches(&host) || port.map(|port| {
                        let url = format!("{}:{}", host, port);
                        self.0.matches(&url)
                    }).unwrap_or(false)
                })
                .unwrap_or(false);
```

File: docs/cue/reference/components.cue
Comment (bruceg) [inline docs/cue/reference/components.cue:726]: I had a comment on a previous commit but it doesn't show up now so I'll repeat it. I am concerned this name doesn't reflect the purpose of this setting. I recommend something like `bypass` or `no_proxy_for`. @jszwedko thoughts?

**Review by bruceg — COMMENTED** [review]
This looks reasonable, but a lot of repetition with possibility for the occasional typo, hence my question about centralizing the configuration.

File: src/config/proxy.rs
Comment (jszwedko) [reply src/config/proxy.rs:7]: I actually think the lower case forms are relatively common albeit they stand out other common environment variable support which prefers capital. These variables were introduced originally in `libwww` with only lower-case support. As other examples, `curl` supports both now, but started with only lowercase; `wget` only supports lowercase.

I like the suggestions regarding standardizing this from Gitlab in https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/ which mentions supporting both cases but preferring lower case.

We should document the support of both cases in Vector though.

File: src/config/proxy.rs
Comment (jszwedko) [reply src/config/proxy.rs:78]: Yeah, that is a definite typo which does seem to indicate a lack of test coverage here. Could we fix the typo and add tests for it?

File: src/kubernetes/client/config/mod.rs
Comment (jszwedko) [reply src/kubernetes/client/config/mod.rs:32]: That's a good question and would help reduce the likelihood that one component would mess it up. Not all sinks use or support proxy configuration though. Do you think there is a way we could do that but still reject proxy config on components that don't allow it?

Alternatively we allow it everywhere, even when it would be unused.

File: docs/cue/reference/components.cue
Comment (jszwedko) [reply docs/cue/reference/components.cue:726]: Yeah, that's a good question. I like `no_proxy` because it matches the commonly used name for this configuration (and the environment variable name), but it is a bit opaque for users that are unfamiliar with `no_proxy` from other contexts.

My inclination is to leave it as-is since `no_proxy` seems to be ubiquitous for proxy configuration.

File: src/kubernetes/client/config/mod.rs
Comment (bruceg) [reply src/kubernetes/client/config/mod.rs:32]: That is the approach I took with `acknowledgements`, in that it is present for all sinks and sources, but ignored (and undocumented) for those that can't use it. If a minority of components do not have support and you want to provide an error, you could add a `ctx.cannot_proxy()?` function call to provide a common error.

File: docs/cue/reference/components.cue
Comment (bruceg) [reply docs/cue/reference/components.cue:726]: Fair enough, if there is precedent then there is no need to change it.

File: scripts/setup_integration/nginx_integration_env.sh
Comment (jdrouet) [reply scripts/setup_integration/nginx_integration_env.sh:42]: it's required to have a proxy available

File: src/kubernetes/client/config/mod.rs
Comment (jdrouet) [reply src/kubernetes/client/config/mod.rs:32]: I completely agree with doing that!
Shouldn't we do the same with `tls`? (and open an issue for that)

Comment (jszwedko) [reply src/kubernetes/client/config/mod.rs]: 👍 it does seem like `tls` and `request` are other good candidates for this treatment. We can follow up on those though.

File: docs/cue/reference/components.cue
Comment (jszwedko) [inline docs/cue/reference/components.cue]:
```suggestion
					The global URL to proxy HTTPS requests through.
					
					If another HTTPS proxy is set in the configuration file or at a component level,
					this one will be overridden.
					
					The lowercase variant has priority over the uppercase one.
```

Typo and adding whitespace.

File: docs/cue/reference/components.cue
Comment (jszwedko) [inline docs/cue/reference/components.cue:969]:
```suggestion
					List of hosts to avoid proxying globally.
					
					Allowed patterns here include:

                     - Domain names. For example, `example.com` will match requests to to `example.com`
                     - Wildcard domains. For example, `.example.com` will match requests to `example.com` and its subdomains
                     - IP addresses. For example, `127.0.0.1` will match requests to 127.0.0.1
                      - CIDR blocks. For example, `192.168.0.0./16` will match requests to any IP addresses in this range
                      - `*` will match all hosts
					
					If another no_proxy value is set in the configuration file or at a component level,
					this one will be overridden.
					
					The lowercase variant has priority over the uppercase one.
```

**Review by jszwedko — COMMENTED** [review]
Nice, this seems cleaner and less error-prone to do it at the top-level to avoid every component merging the proxy config itself.

File: src/kubernetes/client/config/mod.rs
Comment (bruceg) [reply src/kubernetes/client/config/mod.rs]: Agreed, I was thinking the same thing when I was making that suggestion. The one (probably small) gotcha with `request` is that it can't be parsed before calling the sink's `build` function as each sink has different defaults to add to the config. Still, we could pass the config data in to the basic context and so remove it from each sink's config.

File: src/config/mod.rs
Comment (bruceg) [inline src/config/mod.rs:234]: It probably doesn't matter to serde, but I would prefer to leave `inner` as the last element in this struct.

File: docs/cue/reference/components.cue
Comment (bruceg) [inline docs/cue/reference/components.cue:728]: I think this should make explicit that this is not just a list of exact host names but that patterns are also allowed, and what form the patterns may take. In the examples, I see both what appear to be suffix matches and wildcards, so that should be documented.

File: src/config/mod.rs
Comment (bruceg) [inline src/config/mod.rs:335]: Ditto ordering

**Review by bruceg — CHANGES_REQUESTED** [review]
This is much cleaner. My only blocking concern at this point would be documenting the `no_proxy` patterns. The other two comments are just nit picks.

File: docs/cue/reference/components.cue
Comment (jszwedko) [reply docs/cue/reference/components.cue:728]: Mmm, yeah, I missed this with my suggestion to update the env var description https://github.com/timberio/vector/pull/8205#discussion_r679183376. I think we can just copy that here (or share it).

**Review by bruceg — APPROVED** [review]

File: docs/cue/reference/components.cue
Comment (binarylogic) [inline docs/cue/reference/components.cue:994]:
```suggestion
					examples: ["localhost,.example.com", "192.168.0.0./16", "*"]
```

File: docs/cue/reference/components.cue
Comment (binarylogic) [inline docs/cue/reference/components.cue:973]: Should this be `https`?

File: docs/cue/reference/components.cue
Comment (binarylogic) [inline docs/cue/reference/components.cue:962]: Can we add a `warning` here that redirecting otherwise insecure traffic to a non-secure proxy is a security risk. Do you think that's helpful?

File: docs/cue/reference/components.cue
Comment (binarylogic) [inline docs/cue/reference/components.cue:949]: Should this be HTTP?

File: docs/cue/reference/components.cue
Comment (binarylogic) [inline docs/cue/reference/components.cue:1001]: I would prefer that we not support the lower case variant. Is this common? None of our other supported env vars accept case variants.

Comment (jszwedko) [reply docs/cue/reference/components.cue:1001]: It is common specifically for these proxy environment variables to be lower-cased. It is a bit anomalous but dates back to their introduction. See https://github.com/timberio/vector/pull/8205#discussion_r678533122 for more conversation.

Comment (binarylogic) [reply docs/cue/reference/components.cue:1001]: Could we just not document it and look for lower case variants of all env vars? It seems like noise to me to add both in the list of supported env vars. At the very least, we should have a `variants` field or something.

(this is not blocking)

Comment (jszwedko) [reply docs/cue/reference/components.cue:1001]: Yeah, or maybe we could just have a note under the uppercase definitions that the lowercase definition is also recognized. I do realize it is a bit of bloat.

Alternatively, maybe we could expand our cue files to allow env var aliases? That could come in handy in other situations too.

(also not blocking)

File: docs/cue/reference/components.cue
Comment (jdrouet) [reply docs/cue/reference/components.cue:973]: I don't think so, a `https` proxy can be done over `http` 🤔

File: docs/cue/reference/components.cue
Comment (jdrouet) [reply docs/cue/reference/components.cue:962]: I guess, people putting in place a proxy are advanced enough to know it 🤔

Comment (jdrouet) [reply docs/cue/reference/components.cue:962]: If you look at the doc from [AWS for example](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-proxy.html) the `HTTPS_PROXY` uses `http` and no `warning` section talks about it...
We can always update it if we see some issues about it.

**Pull request merged**
